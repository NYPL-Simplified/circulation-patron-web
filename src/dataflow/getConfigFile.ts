import { AppSetupError } from "./../errors";
import { readFileSync, existsSync } from "fs";
import path from "path";
import { AppConfig } from "interfaces";
import YAML from "yaml";

/**
 * Reads a config file either from local path or
 * http request, parses it, and returns it as an object
 */
export default async function getAppConfig(
  configFileSetting: string
): Promise<AppConfig> {
  if (configFileSetting.startsWith("http")) {
    return await fetchConfigFile(configFileSetting);
  }
  const configFilePath = path.join(process.cwd(), configFileSetting);
  if (!existsSync(configFilePath)) {
    throw new AppSetupError("Config file not found at: " + configFilePath);
  }
  const text = readFileSync(configFilePath, "utf8");
  return parseConfigText(text);
}

/**
 * Fetches a config file from the network, parses it into
 * an object and returns it
 */
async function fetchConfigFile(configFileUrl: string): Promise<AppConfig> {
  try {
    const response = await fetch(configFileUrl);
    const text = await response.text();
    const parsed = parseConfigText(text);
    return parsed;
  } catch (e) {
    throw new Error("Could not fetch config file at: " + configFileUrl);
  }
}

/**
 * Parses the raw text of a config file into an object.
 */
function parseConfigText(raw: string): AppConfig {
  const config = YAML.parse(raw);

  // specifically set defaults for a couple values.
  const companionApp =
    config.companion_app === "openebooks" ? "openebooks" : "simplye";
  const axisNowDecrypt = config.axisnow_decrypt === true;

  // otherwise assume the file is properly structured.
  return {
    ...config,
    axisNowDecrypt,
    companionApp
  };
}
