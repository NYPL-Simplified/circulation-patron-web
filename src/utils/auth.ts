import {
  AuthProvider,
  AuthMethod,
  BasicAuthMethod
} from "opds-web-client/lib/interfaces";

export const BASIC_AUTH_ID = "http://opds-spec.org/auth/basic";

/**
 * This is a typescript type guard used to narrow the type of the
 * passed in provider.
 */
function isBasicAuthProvider(
  provider: AuthProvider<AuthMethod | BasicAuthMethod>
): provider is AuthProvider<BasicAuthMethod> {
  const correctId = provider.id === BASIC_AUTH_ID;
  // const hasLabels = !!(provider.method as BasicAuthMethod).labels;
  const hasLabels = "labels" in provider.method;
  return correctId && hasLabels;
}

export function getBasicAuthProvider(
  providers: AuthProvider<AuthMethod>[] | null
): AuthProvider<BasicAuthMethod> | undefined {
  if (!providers) return undefined;
  for (const provider of providers) {
    if (isBasicAuthProvider(provider)) {
      return provider;
    }
  }
  return undefined;
}
