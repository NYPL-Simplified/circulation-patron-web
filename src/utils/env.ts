/* eslint-disable no-underscore-dangle */
/**
 * Simply exporting processed env vars
 */

export const CONFIG_FILE = process.env.CONFIG_FILE;
export const REACT_AXE = process.env.REACT_AXE;
export const IS_DEVELOPMENT = process.env.NODE_ENV === "development";
export const IS_SERVER = typeof window === "undefined";

// where the redux store is kept on window
export const __NEXT_REDUX_STORE__ = "__NEXT_REDUX_STORE__";
// where we store the library data
export const __LIBRARY_DATA__ = "__LIBRARY_DATA__";

export const BUGSNAG_API_KEY = process.env.NEXT_PUBLIC_BUGSNAG_API_KEY;
export const APP_VERSION = process.env.APP_VERSION;
