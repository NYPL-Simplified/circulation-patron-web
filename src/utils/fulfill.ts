import { MediaLink, FulfillmentLink, OPDS1 } from "interfaces";

type Download = {
  type: "download";
  url: string;
  filename: string;
  extension: string;
  buttonLabel: string;
};
type ReadInternal = {
  type: "read-internal";
  url: string;
};
type ReadExternal = {
  type: "read-external";
  url: string;
  buttonLabel: string;
};
type Unsupported = {
  type: "unsupported";
};

/**
 * This is how the user perceives the fulfillment, not necessarily how
 * we perform the fulfillment.
 */
type FulfillmentType = Download | ReadInternal | ReadExternal | Unsupported;

/**
 * Mapping links to fulfillment protocols based on
 * https://docs.google.com/document/d/1dli5mgTbVaURN_B2AtUmPhgpaFUVqOqrzsaoFvCXnkA/edit?pli=1#
 */
export function getFulfillmentType(
  link: MediaLink | FulfillmentLink
): FulfillmentType {
  const endContentType = fixMimeType(link.type);
  const indirectType = "indirectType" in link ? link.indirectType : undefined;

  // there is no support for books with "Libby" DRM
  if (
    endContentType === OPDS1.OverdriveEbookMediaType ||
    indirectType === OPDS1.OverdriveEbookMediaType
  ) {
    return { type: "unsupported" };
  }
  switch (endContentType) {
    case OPDS1.PdfMediaType:
    case OPDS1.Mobi8Mediatype:
    case OPDS1.MobiPocketMediaType:
    case OPDS1.EpubMediaType:
      return {
        url: link.url,
        type: "download",
        buttonLabel: `Download`,
        extension: `blah`,
        filename: "filename"
      };

    case OPDS1.AtomMediaType:
      return {
        url: link.url,
        type: "read-external",
        buttonLabel: "Read on Overdrive"
      };

    case OPDS1.AxisNowWebpubMediaType:
      return {
        // here parse this into the actual url we need to send the user to
        url: `/read/${encodeURIComponent(link.url)}`,
        type: "read-internal"
      };
  }
  return {
    type: "unsupported"
  };
}

export function fixMimeType(
  mimeType: OPDS1.AnyBookMediaType | "vnd.adobe/adept+xml"
): OPDS1.AnyBookMediaType {
  return mimeType === "vnd.adobe/adept+xml"
    ? "application/vnd.adobe.adept+xml"
    : mimeType;
}
