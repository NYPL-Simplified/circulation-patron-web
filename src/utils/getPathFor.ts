import { PathFor } from "opds-web-client/lib/interfaces";
import encodeUrlParam from "utils/url";

const getPathFor = (librarySlug: string | null): PathFor => (
  collectionUrl?: string | null,
  bookUrl?: string | null
) => {
  let path = "";
  if (librarySlug) {
    path += "/" + librarySlug;
  }
  if (collectionUrl) {
    const preparedCollectionUrl = encodeUrlParam(collectionUrl);
    if (preparedCollectionUrl) {
      path += `/collection/${preparedCollectionUrl}`;
    }
  }
  if (bookUrl) {
    path += `/book/${encodeUrlParam(bookUrl)}`;
  }
  if (!path) {
    path = "/";
  }
  return path;
};

export default getPathFor;
