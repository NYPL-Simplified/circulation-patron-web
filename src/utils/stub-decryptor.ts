/**
 * This is a file that can be required by apps that don't have access to the axisnow decryptor
 */

throw new Error(
  "This application mistakenly required the AxisNow Decryptor, when it does not have access"
);

export default {};
