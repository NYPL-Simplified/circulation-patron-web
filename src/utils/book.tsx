import * as React from "react";
import {
  AnyBook,
  BookData,
  BookMedium,
  BorrowableBook,
  FulfillableBook,
  OnHoldBook,
  ReservableBook,
  ReservedBook
} from "interfaces";
import { BookFulfillmentState } from "interfaces";
import { Book, Headset } from "../icons";
import { DEFAULT_AVAILABILITY } from "utils/constants";

export function getAuthors(book: BookData, lim?: number): string[] {
  // select contributors if the authors array is undefined or empty.
  const allAuth =
    typeof book.authors?.length === "number" && book.authors.length > 0
      ? book.authors
      : typeof book.contributors?.length === "number" &&
        book.contributors.length > 0
      ? book.contributors
      : ["Authors unknown"];

  // now limit it to however many
  if (lim) {
    return allAuth.slice(0, lim);
  }
  return allAuth;
}

export function availabilityString(book: BookData) {
  if (book.openAccessLinks && book.openAccessLinks.length > 0)
    return "This open-access book is available to keep forever.";

  const availableCopies = book.copies?.available;
  const totalCopies = book.copies?.total;
  return typeof availableCopies === "number" && typeof totalCopies === "number"
    ? `${availableCopies} out of ${totalCopies} copies available.`
    : "Number of books available is unknown.";
}

export function queueString(book: BookData) {
  const holds = book.holds?.total;
  return typeof holds === "number"
    ? `There are ${holds} other patrons in the queue.`
    : "";
}

export function bookIsFulfillable(book: AnyBook): book is FulfillableBook {
  return book.status === "fulfillable";
}

export function bookIsReserved(book: AnyBook): book is ReservedBook {
  return book.status === "reserved";
}

export function bookIsReservable(book: AnyBook): book is ReservableBook {
  return book.status === "reservable";
}

export function bookIsOnHold(book: AnyBook): book is OnHoldBook {
  return book.status === "on-hold";
}

export function bookIsBorrowable(book: AnyBook): book is BorrowableBook {
  return book.status === "borrowable";
}

export function bookIsAudiobook(book: BookData): boolean {
  if (getMedium(book) === "http://bib.schema.org/Audiobook") {
    return true;
  }
  return false;
}

export const bookMediumMap: {
  [key in BookMedium]: {
    name: string;
    icon: React.ComponentType<{ className?: string }>;
  };
} = {
  "http://bib.schema.org/Audiobook": {
    name: "Audiobook",
    icon: Headset
  },
  "http://schema.org/EBook": { name: "eBook", icon: Book },
  "http://schema.org/Book": { name: "Book", icon: Book }
};

export function getMedium(book: BookData): BookMedium | "" {
  if (!book.raw || !book.raw["$"] || !book.raw["$"]["schema:additionalType"]) {
    return "";
  }

  return book.raw["$"]["schema:additionalType"].value
    ? book.raw["$"]["schema:additionalType"].value
    : "";
}
