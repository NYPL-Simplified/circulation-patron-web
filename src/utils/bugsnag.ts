import * as React from "react";
import Bugsnag from "@bugsnag/js";
import BugsnagPluginReact from "@bugsnag/plugin-react";
import { APP_VERSION, BUGSNAG_API_KEY } from "utils/env";

if (BUGSNAG_API_KEY) {
  Bugsnag.start({
    apiKey: BUGSNAG_API_KEY,
    appVersion: APP_VERSION,
    plugins: [new BugsnagPluginReact()]
  });
}

export const BugsnagErrorBoundary = BUGSNAG_API_KEY
  ? Bugsnag.getPlugin("react")?.createErrorBoundary(React)
  : undefined;

export default Bugsnag;
