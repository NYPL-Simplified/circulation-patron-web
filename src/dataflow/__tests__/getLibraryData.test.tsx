import {
  fetchCatalog,
  getCatalogUrl,
  fetchAuthDocument,
  getLibraryData,
  getLibrarySlugs
} from "../getLibraryData";
import getConfigFile from "../getConfigFile";
import fetchMock from "jest-fetch-mock";
import ApplicationError, { PageNotFoundError } from "errors";
import * as env from "utils/env";
import rawCatalog from "test-utils/fixtures/opds-feed";

function setEnv({
  CONFIG_FILE = undefined,
  CIRCULATION_MANAGER_BASE = undefined,
  REGISTRY_BASE = undefined
}: any) {
  (env.CONFIG_FILE as any) = CONFIG_FILE;
  (env.CIRCULATION_MANAGER_BASE as any) = CIRCULATION_MANAGER_BASE;
  (env.REGISTRY_BASE as any) = REGISTRY_BASE;
}

describe("fetchCatalog", () => {
  test("calls fetch with catalog url", async () => {
    fetchMock.mockResponseOnce(rawCatalog);
    const promise = await fetchCatalog("some-url");
    // expect();
  });

  test("Throws error if catalog is not correct format", async () => {
    fetchMock.mockResponseOnce("something invalid");
    const promise = fetchCatalog("a url somewhere");
    await expect(promise).rejects.toThrowError(ApplicationError);
    await expect(promise).rejects.toThrowErrorMatchingSnapshot();
  });

  test("Throws error if fetch fails", async () => {
    fetchMock.mockRejectOnce();
    const promise = fetchCatalog("not a valid url");
    await expect(promise).rejects.toThrowError(ApplicationError);
    await expect(promise).rejects.toThrowErrorMatchingInlineSnapshot(
      `"Could not fetch catalog at not a valid url"`
    );
  });
});

jest.mock("../getConfigFile");
const mockGetConfigFile = getConfigFile as jest.MockedFunction<
  typeof getConfigFile
>;
mockGetConfigFile.mockResolvedValue({
  somelibrary: "somelibraryurl",
  anotherlib: "anotherliburl"
});

describe("getCatalogUrl", () => {
  test("throws error if there is a library slug and CIRCULATION_MANAGER_BASE", async () => {
    setEnv({ CIRCULATION_MANAGER_BASE: "some-base" });
    const promise = getCatalogUrl("some-slug");
    await expect(promise).rejects.toThrowError(ApplicationError);
    await expect(promise).rejects.toThrowErrorMatchingInlineSnapshot(
      `"App is running with a single Circ Manager, but you're trying to access a multi-library route: some-slug"`
    );
  });

  test("throws PageNotFoundError if running multiple libraries and no slug provided", async () => {
    setEnv({
      CONFIG_FILE: "config-file"
    });
    const promise = getCatalogUrl();
    await expect(promise).rejects.toThrowError(PageNotFoundError);
    await expect(promise).rejects.toThrowErrorMatchingInlineSnapshot(
      `"Library slug must be provided when running with multiple libraries."`
    );
  });

  test("throws PageNotFoundError if no entry found in config file for library", async () => {
    setEnv({
      CONFIG_FILE: "config-file"
    });
    const promise = getCatalogUrl("not there slug");
    await expect(promise).rejects.toThrowError(PageNotFoundError);
    await expect(promise).rejects.toMatchInlineSnapshot(
      `[Page Not Found Error: No CONFIG_FILE entry for library: not there slug]`
    );
  });

  test("returns url for existing library in config file", async () => {
    setEnv({ CONFIG_FILE: "config-file" });
    const promise = getCatalogUrl("anotherlib");
    await expect(promise).resolves.toBe("anotherliburl");
  });

  test("Works for Registry Base", async () => {
    setEnv({ REGISTRY_BASE: "reg-base" });
    const url = await getCatalogUrl("lib-on-registry");
    expect(url).toBe("http://lib-on-registry");
  });
});
