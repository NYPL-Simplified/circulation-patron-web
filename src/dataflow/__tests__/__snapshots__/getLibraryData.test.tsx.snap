// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchCatalog Throws error if catalog is not correct format 1`] = `
"Could not fetch catalog at a url somewhere
Base Error:
Non-whitespace before first tag.
Line: 0
Column: 1
Char: s"
`;

exports[`fetchCatalog Throws error if catalog is not correct format: "Could not fetch catalog at a url somewhereNon-whitespace before first tag." 1`] = `
"Could not fetch catalog at a url somewhere
Base Error:
Non-whitespace before first tag.
Line: 0
Column: 1
Char: s"
`;
